xapian-core/weight/tfidfweight.cc

```diff
@@ -138,19 +197,21 @@ TfIdfWeight::unserialise(const string & s) const
     const char *end = ptr + s.size();
     double slope = unserialise_double(&ptr, end);
     double delta = unserialise_double(&ptr, end);
-    string normals(ptr, end);
-    ptr += 3;
+    wdf_norm wdf_normalization = static_cast<wdf_norm>(*(ptr)++);
+    idf_norm idf_normalization = static_cast<idf_norm>(*(ptr)++);
+    wt_norm wt_normalization = static_cast<wt_norm>(*(ptr)++);
```

Comment:
This doesn't check that there's actually enough data in the string supplied (the existing code isn't quite right either).  (The first byte read is OK as there's guaranteed to be a zero byte on the end (since C++11), but the others are off the end of the string, so undefined behaviour).

These serialisations are passed across the network, so we need to take extra care not to assume they are valid.

Instead of throwing an exception if `ptr != end` after we process these bytes, we can instead throw if `end - ptr != 3` before we process them - that way if the string is too short we throw an exception instead of invoking undefined behaviour.

Reply:
Updated